R/namespace_linter.R

```diff
@@ -33,97 +33,124 @@ namespace_linter <- function(check_exports = TRUE, check_nonexports = TRUE) {
+  exports <- getNamespaceExports(namespace)
+  lazydata <- names(.getNamespaceInfo(namespace, "lazydata"))
+  symbol %in% c(exports, lazydata)
+}
+
+build_unexported_lint <- function(symbol, symbol_node, package, source_expression) {
+  line1 <- as.integer(xml2::xml_attr(symbol_node, "line1"))
+  col1 <- as.integer(xml2::xml_attr(symbol_node, "col1"))
+  col2 <- as.integer(xml2::xml_attr(symbol_node, "col2"))
+  Lint(
+    filename = source_expression$filename,
+    line_number = line1,
+    column_number = col1,
+    type = "warning",
+    message = sprintf("'%s' is not exported from {%s}.", symbol, package),
+    line = source_expression$file_lines[[line1]],
+    ranges = list(c(col1, col2))
+  )
+}
+
+build_unnecessary_private_lint <- function(symbol, symbol_node, package, source_expression) {
+  line1 <- as.integer(xml2::xml_attr(symbol_node, "line1"))
+  col1 <- as.integer(xml2::xml_attr(symbol_node, "col1"))
+  col2 <- as.integer(xml2::xml_attr(symbol_node, "col2"))
+  Lint(
+    filename = source_expression$filename,
+    line_number = line1,
+    column_number = col1,
+    type = "style",
+    message = sprintf("'%s' is exported from {%s}. Use %s::%s instead.", symbol, package, package, symbol),
+    line = source_expression$file_lines[[line1]],
+    ranges = list(c(col1, col2))
+  )
+}
+
+build_unfound_symbol_lint <- function(symbol, symbol_node, package, source_expression) {
+  line1 <- as.integer(xml2::xml_attr(symbol_node, "line1"))
+  col1 <- as.integer(xml2::xml_attr(symbol_node, "col1"))
+  col2 <- as.integer(xml2::xml_attr(symbol_node, "col2"))
+  Lint(
+    filename = source_expression$filename,
+    line_number = line1,
+    column_number = col1,
+    type = "warning",
+    message = sprintf("'%s' does not exist in {%s}.", symbol, package),
+    line = source_expression$file_lines[[line1]],
+    ranges = list(c(col1, col2))
+  )
+}
+
+build_non_namespace_lint <- function(package_node, message, source_expression) {
+  line1 <- as.integer(xml2::xml_attr(package_node, "line1"))
+  col1 <- as.integer(xml2::xml_attr(package_node, "col1"))
+  col2 <- as.integer(xml2::xml_attr(package_node, "col2"))
+  Lint(
+    filename = source_expression$filename,
+    line_number = line1,
+    column_number = col1,
+    type = "warning",
+    message = message,
+    line = source_expression$file_lines[[line1]],
+    ranges = list(c(col1, col2))
+  )
+}
+
+build_uninstalled_lint <- function(package, package_node, source_expression) {
+  line1 <- as.integer(xml2::xml_attr(package_node, "line1"))
+  col1 <- as.integer(xml2::xml_attr(package_node, "col1"))
+  col2 <- as.integer(xml2::xml_attr(package_node, "col2"))
+  Lint(
+    filename = source_expression$filename,
+    line_number = line1,
+    column_number = col1,
+    type = "warning",
+    message = sprintf("Package '%s' is not installed.", package),
+    line = source_expression$file_lines[[line1]],
+    ranges = list(c(col1, col2))
+  )
+}
```

Comment:
Aren't all these special cases of `xml_nodes_to_lint()`?

Reply:
i think so... will need the vectorized messages though in the current approach